### ISSUE TITLE
Does not support "enum" schemas

### ISSUE THREAD

**Issue by Venryx** [thenativeweb/get-graphql-from-jsonschema#326]
The "enum" schema is pretty simple and commonly used, so it seems a good candidate for inclusion.

For reference, here is how the other json-schema-to-graphql-schema library does it: https://github.com/HerbCaudill/jsonschema2graphql/blob/master/src/schemaReducer.ts#L91

I don't have time at the moment, but would you be open for a PR for this in the future?

**Comment by Venryx** [comment]
Actually, I went ahead and added support for "enum" in my fork: https://github.com/thenativeweb/get-graphql-from-jsonschema/compare/main...Venryx:main#diff-ea60dc23af8cbb810ebcab7c5c9a9974cf4bf77224e516c08028005333695be6

Let me know what I would need to change for a pull-request to be acceptable.

**Comment by Venryx** [comment]
I was hitting some annoyances, and one big stopper, from the fact that this library doesn't yet support the "$ref" field in schemas, so I went ahead and added basic support for it.

Usage:
```
// At some point (can be before, after, or in the same loop), add the graphql types referenced by the "$ref" fields. Example:
const schemasUsedForRefs = [...];
for (const schema of schemasUsedForRefs) {
  const {typeDefinitions} = getGraphqlSchemaFromJsonSchema({ schema.$id, schema });
  addTypesToGraphQL(typeDefinitions);
}

// For schemas with "$ref" fields, pass a "refToTypeName" function during graphql type generation.
// This allows the library to resolve the "$ref" names to graphql type names. (will error otherwise)
const schemasThatUseRefs = [...];
for (const schema of schemasThatUseRefs) {
  const {typeDefinitions} = getGraphqlSchemaFromJsonSchema({
    schema.$id,
    schema,
    refToTypeName: refName=>{
      // Can add custom logic here, depending on the naming used for generating the schemas' graphql types.
      return `${refName}T0`;
    },
  });
  addTypesToGraphQL(typeDefinitions);
}
```

And here are the code changes made, for adding support for both `enum` and `$ref`: https://github.com/thenativeweb/get-graphql-from-jsonschema/compare/main...Venryx:a9bb3607851538dc4d209acd8f3f5060e5a9dbda#diff-ea60dc23af8cbb810ebcab7c5c9a9974cf4bf77224e516c08028005333695be6

Again, please let me know what changes you'd want me to make for it to be acceptable as a pull-request.

**Comment by yeldiRium** [comment]
Hi @Venryx :wave: 

Thank you for raising this issue and taking the time to implement a solution.

The support for the `enum` field we'd gladly accept, but I do not want to add untested features. Do you mind adding some appropriate unit tests? And have you used this already in a GraphQL server? I don't know how JavaScript implementations of GraphQL handle enums and I would not assume that GraphQL-enum-representations in JavaScript can be easily validated using a JsonSchema `enum` field without testing it. [The GraphQL schema documentation theorizes that enums in JavaScript might be implemented as integers](https://graphql.org/learn/schema/#enumeration-types), which would definitely be incompatible to JsonSchemas that use an `enum` field for e.g. strings.

As for the `$ref` support: I like your approach to completely avoid the issue of loading referenced schemas and having users convert their schemas first and then providing a way to resolve them. It feels pretty rough though and won't work for circular references. I'll have to think about this and get back to you.

So I suggest we split the two topics. If you want, you can go ahead and open a PR with the `enum` support. I'll get back to you regarding the `$ref` support later.

Does that sound good to you?
Kind regards,
yeldiR

**Comment by Venryx** [comment]
> The support for the enum field we'd gladly accept, but I do not want to add untested features. Do you mind adding some appropriate unit tests?

Yes, I'll look into that. I can't say exactly when, but it's on my list. (page bookmarked)

> And have you used this already in a GraphQL server?

Yes, I'm using it currently for this project: https://github.com/debate-map/app

However, I am still in the middle of a server refactor, so I have not been able to "thoroughly test" usage of the enum types yet. I haven't hit issues so far though.

> The GraphQL schema documentation theorizes that enums in JavaScript might be implemented as integers, which would definitely be incompatible to JsonSchemas that use an enum field for e.g. strings.

Yes, I'll need to look more closely at how to handle enums with varying string/number values. In my project, I only use string-based TypeScript enums (eg. `enum Fruit { apple = "apple", pear = "pear" }`), so I will definitely need to add tests and such to make sure it's usable for number-based ones.

> As for the $ref support: I like your approach to completely avoid the issue of loading referenced schemas and having users convert their schemas first and then providing a way to resolve them. It feels pretty rough though and won't work for circular references. I'll have to think about this and get back to you.

Why would it not work for circular references? The library does not actually "resolve" the graphql types that are referenced, it merely provides a way for the user to say "Okay, here is the graphql type name that we'll be using for this json-schema name; I'll provide it to my GraphQL server myself, and you can just trust that it'll be there at runtime."

Because of this approach, circular references would just work, no? (ie. both "groups" can reference types within the other because the developer provides the formula, and this can be based on anything -- whether that be hard-coding, searching, or what I currently do, of simply knowing the naming scheme that is used for the graphql-type generation and adding that name-transformation to the `refToTypeName` function)

> So I suggest we split the two topics. If you want, you can go ahead and open a PR with the enum support. I'll get back to you regarding the $ref support later.

Yes, that sounds good. I'll make a new issue for `$ref` support, and make separate pull-requests for them later.
